### PR TITLE
zlib-ng patch for WD suppresion fixes

### PR DESCRIPTION
**Pull request by carlossanlop**
Follow up of https://github.com/dotnet/runtime/pull/105433

Add patch that include the zlib-ng changes.

I also made a couple of minor changes to the code to make sure it matches both the patch and the upstream PR.

Upstream PR: https://github.com/zlib-ng/zlib-ng/pull/1762

### CODE REVIEW

**Comment by jkotas** [comment]
What's the value of the checked in patch? Why is not good enough to have a link to a commit to apply?

File: src/native/external/zlib-ng/deflate_rle.c [hunk 1]
```diff
@@ -58,9 +58,8 @@ Z_INTERNAL block_state deflate_rle(deflate_state *s, int flush) {
 
         /* Emit match if have run of STD_MIN_MATCH or longer, else emit literal */
         if (match_len >= STD_MIN_MATCH) {
-            Assert(s->strstart <= UINT16_MAX, "strstart should fit in uint16_t");
-            Assert(s->match_start <= UINT16_MAX, "match_start should fit in uint16_t");
-            check_match(s, (Pos)s->strstart, (Pos)s->strstart - 1, match_len);
+            Assert((s->strstart - 1) <= UINT16_MAX, "strstart should fit in uint16_t");
```
Comment (carlossanlop) [inline src/native/external/zlib-ng/deflate_rle.c:61]:
```suggestion
            Assert(s->strstart  <= UINT16_MAX, "strstart should fit in uint16_t");
```

File: src/native/external/patches/zlib-ng/0001-zlib-ng-avoid-suppressing-WD4242-and-WD4244-105433.patch [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+diff --git a/src/native/external/zlib-ng/deflate_rle.c b/src/native/external/zlib-ng/deflate_rle.c
+index cd085099460..b23a275c426 100644
+--- a/src/native/external/zlib-ng/deflate_rle.c
++++ b/src/native/external/zlib-ng/deflate_rle.c
+@@ -58,7 +58,8 @@ Z_INTERNAL block_state deflate_rle(deflate_state *s, int flush) {
+ 
+         /* Emit match if have run of STD_MIN_MATCH or longer, else emit literal */
+         if (match_len >= STD_MIN_MATCH) {
+-            check_match(s, s->strstart, s->strstart - 1, match_len);
++            Assert((s->strstart - 1) <= UINT16_MAX, "strstart-1 should fit in uint16_t");
```
Comment (carlossanlop) [inline src/native/external/patches/zlib-ng/0001-zlib-ng-avoid-suppressing-WD4242-and-WD4244-105433.patch:209]:
```suggestion
+            Assert(s->strstart <= UINT16_MAX, "strstart should fit in uint16_t");
```

**Comment by carlossanlop** [comment]
> What's the value of the checked in patch? Why is it not good enough to have a link to a commit(s) to apply?

I noticed that if I go to https://github.com/dotnet/runtime/pull/105433.patch , the patch that gets generated contains the whole commit history from the PR, not just the final commit. That's unnecessary noise.

This proposed patch on the other hand was generated out of main, from the commit after I merged the PR.

It also contains the nits that I missed in the PR (and that AI am additionally fixing here).

**Comment by jkotas** [comment]
> I noticed that if I go to https://github.com/dotnet/runtime/pull/105433.patch , the patch that gets generated contains the whole commit history from the PR, not just the final commit. 

If you do not like that, you can replace the link to the patch with the link to the commit: `https://github.com/dotnet/runtime/commit/ecdb625035e0e3fb7c51e908713d96d2cb2080c8`

We do not maintain the patch files for any other vendored libraries. We link to commits or PRs that provide the same information:
- https://github.com/dotnet/runtime/blob/main/src/native/external/libunwind-version.txt
- https://github.com/dotnet/runtime/blob/main/src/native/external/llvm-libunwind-version.txt

> It also contains the nits that I missed in the PR

I would not bother with nits unless they are actual bugs. I expect you will get a bunch of feedback on the upstream PR and the upstreamed change will be different from our private patch. That's fine. We will pick it up during the next libng version bump.

**Comment by carlossanlop** [comment]
I guess we can close this then. It's going to be temporary anyway, since we will eventually update to the zlib-ng release containing the final fix.

Just wanted to note that we did have patches for zlib and zlib-intel before, but I deleted them yesterday alongside the source code of zlib and zlib-intel. Here: https://github.com/dotnet/runtime/tree/release/9.0-preview7/src/native/external/patches

**Pull request closed without merging**